include/seahorn/Expr/TypeChecker.hh

```diff
@@ -1,17 +1,168 @@
 #pragma once
-
-#include "seahorn/Expr/ExprApi.hh"
+#include "seahorn/Expr/ExprCore.hh"
+#include "seahorn/Expr/ExprOpSort.hh"
 
 namespace expr {
 class TypeChecker {
   class Impl;
-  Impl* m_impl;
+  Impl *m_impl;
 
 public:
   TypeChecker();
   ~TypeChecker();
   Expr sortOf(Expr e) { return this->typeOf(e); }
-  Expr typeOf(Expr e); 
+  Expr typeOf(Expr e);
+  Expr getErrorExp();         // to be called after typeOf() or sortOf()
+  ExprSet getEndExps(Expr e); // to be called after typeOf() or sortOf()
 };
 
+namespace op {
+namespace typeCheck {
+enum Comparison { Equal, GreaterEqual };
+
+template <Comparison compareType, unsigned int numChildren>
+static inline bool checkNumChildren(Expr exp) {
```

Comment:
the way you use this functions, it seems unnatural to pass `numChildren` as a template parameter. This increases compilation time without much gain in performance. Might as well pass it as an argument.

Probably this is because I've suggested it. But I had a different -- more template based -- design in mind where this would make more sense.